jni/src/main/java/com/horizen/merkletreenative/ZenBoxMerkleTree.java

```diff
@@ -0,0 +1,159 @@
+package com.horizen.merkletreenative;
+
+import com.horizen.librustsidechains.FieldElement;
+import com.horizen.librustsidechains.Library;
+
+public class ZenBoxMerkleTree implements AutoCloseable {
+
+    // Loading the Rust library which contains all the underlying logic
+    static {
+        Library.load();
+    }
+
+    private long merkleTreePointer;
+
+    private void checkPointer() throws IllegalStateException {
+        if (merkleTreePointer == 0)
+            throw new IllegalStateException("ZenBoxMerkleTree instance was freed");
+    }
+
+    // Constructor is intended to be called from inside of the Rust environment getting a raw pointer to ZenBoxSMT Rust-instance
+    private ZenBoxMerkleTree(long merkleTreePointer) {
+        this.merkleTreePointer = merkleTreePointer;
+    }
+
+    // Gates to the Rust-side API
+    private static native ZenBoxMerkleTree nativeInit(int height, String dbPath);
+    private static native void nativeFreeAndDestroy(long merkleTreePointer);
+    private static native void nativeFree(long merkleTreePointer);
+    private static native long nativeResetBitvector(long merkleTreePointer);
+    private static native long nativeGetAbsolutePosition(FieldElement leaf, int height);
+
+    private native long nativeGetPosition(FieldElement leaf);
+    private native boolean nativeIsPositionEmpty(long position);
+    private native boolean nativeIsBoxSpent(long position);
+    private native void nativeAddBox(FieldElement leaf, long position);
+    private native void nativeRemoveBox(long position);
+    private native FieldElement nativeGetBox(long position);
+    private native FieldElement nativeStateRoot();
+    private native FieldElement nativeBitvectorRoot();
+    private native MerklePath nativeGetStateMerklePath(long leafPosition);
+    private native MerklePath nativeGetBitvectorMerklePath(long leafPosition);
+    private native void nativeFlush();
+
+    // Creates ZenBoxSMT Rust-instance, which though can be not created in case of underlying errors
+    // The returned ZenBoxMerkleTree object should be checked with the 'isInitialized' method
+    // Note: ZenBoxSMT Rust-instance may be not created due to errors such as incorrectly specified tree height or DB path, insufficient disk space and so on.
+    //       In such a case constructor will be called with merkleTreePointer = null so ZenBoxMerkleTree will be empty.
+    //       Thus ZenBoxMerkleTree should be checked with 'isInitialized' method before any usage
```

Comment:
Nice, but in the future this won't be the way we deal with it.
Look at this opened issue:

https://github.com/HorizenOfficial/zendoo-sc-cryptolib/issues/31

Reply:
Yes, indeed it would be better to use exceptions to provide a more informative report about what happened inside of the Rust methods in case of some runtime errors